core/src/main/scala/org/apache/spark/shuffle/ShuffleBlockPusher.scala

```diff
@@ -327,10 +338,32 @@ private[spark] class ShuffleBlockPusher(conf: SparkConf) extends Logging {
         s"stop.")
       return false
     } else {
+      if (reqsInFlight <= 0 && pushRequests.isEmpty && deferredPushRequests.isEmpty) {
```

Comment:
Note for myself: Need to add a test for this particular condition

Reply:
I tried adding test for this particular case, but due to `shuffleClient.pushBlocks` is mocked, it becomes synchronous and hard to test this case. May be we need to have a `FakeShuffleClient` and then have the async behavior to test this case. Any other suggestions?

Reply:
Tried waiting on the latch for one of the shuffleClient.pushBlocks call but this is still not working. Need to have a latch on the notifyDriverAboutPushCompletion so that the tests won't complete before all the blocks are pushed. Also we need to have the mock of when(shuffleClient.pushBlocks to be running in a separate thread without that it is hard to test the problematic behavior. So instead just checking assert(bytesInFlight <= 0) in notifyDriverAboutPushCompletion itself. Let me know if you have other thoughts.

Reply:
Okay I addressed the issues here, now the test should be good.